modelci/runs.py

```diff
@@ -0,0 +1,308 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+"""
+Author: USER
+Email: yli056@e.ntu.edu.sg
+Date: 10/2/2020
+"""
+import os
+import random
+import signal
+# NOQA: B404
+import subprocess
+import sys
+from pathlib import Path
+
+import docker
+
+from modelci.app.config import SERVER_HOST, SERVER_PORT
+from modelci.config import (
+    MONGO_PORT, MONGO_USERNAME, MONGO_PASSWORD, MONGO_DB, MONGO_CONTAINER_LABEL,
+    CADVISOR_CONTAINER_LABEL, CADVISOR_PORT,
+    DCGM_EXPORTER_CONTAINER_LABEL, GPU_METRICS_EXPORTER_CONTAINER_LABEL, NODE_EXPORTER_PORT,
+)
+from modelci.utils import Logger
+from modelci.utils.docker_api_utils import list_containers, get_image, check_container_status
+
+logger = Logger(__name__, welcome=False)
+
+backend_process = None
+
+
+def start():
+    """Start the ModelCI service."""
+    docker_client = docker.from_env()
+    download_serving_containers(docker_client)
+
+    if list_containers(docker_client, filters={'label': MONGO_CONTAINER_LABEL, 'status': 'running'}):
+        logger.info('MongoDB already started.')
+    else:
+        start_mongodb(docker_client)
+
+    if list_containers(docker_client, filters={'label': CADVISOR_CONTAINER_LABEL, 'status': 'running'}):
+        logger.info('cAdvisor already started.')
+    else:
+        start_cadvisor(docker_client)
+
+    if list_containers(docker_client, filters={'label': DCGM_EXPORTER_CONTAINER_LABEL, 'status': 'running'}) and \
+            list_containers(
+                docker_client,
+                filters={'label': GPU_METRICS_EXPORTER_CONTAINER_LABEL, 'status': 'running'}
+            ):
+        logger.info('Node exporter already started.')
+    else:
+        start_node_exporter(docker_client)
+
+    start_fastapi_backend()
+
+
+def stop():
+    """Stop the ModelCI service"""
+    docker_client = docker.from_env()
+    stop_mongodb(docker_client)
+    stop_cadvisor(docker_client)
+    stop_node_exporter(docker_client)
+    stop_fastapi_backend()
+
+
+def _check_service_started(docker_client, label):
+    containers = list_containers(docker_client, filters={'label': [label]})
+    if containers:
+        name = containers[0].name
+        # container exist
+        if check_container_status(docker_client, name):
+            # container is running
+            logger.info(f'Already running, found container name={name}.')
+        else:
+            containers[0].start()
+            logger.info(f'Already exist, found container name={name}. Service started.')
+        return name
+
+
+def _check_process_running(pid_file):
+    if Path(pid_file).exists():
+        with open(pid_file) as f:
+            pid = int(f.read())
+        try:
+            os.killpg(os.getpgid(pid), signal.ITIMER_REAL)
+            logger.info(f'Already running, found PID={pid}.')
+            return pid
+        except OSError:
+            pass
+    return False
+
+
+def _stop_service(docker_client, label):
```

Comment:
what does '-' prefix mean in function name? Is this function private to other classes or files? if not, please remove them.

Reply:
A private method

Reply:
you should build a class so we can call that private. According to my comment above.